libresin/libresin/core/resources/shader_resource.cpp

```diff
@@ -0,0 +1,202 @@
+  auto path    = std::filesystem::path{arg_val};
+  if (path.empty()) {
+    helpers::log_sh_err(curr_line_num, "The include macro argument cannot be empty.");
+    return std::nullopt;
+  }
+
+  if (path.is_absolute()) {
+    helpers::log_sh_err(curr_line_num, "The include macro argument cannot be an absolute path.");
+    return std::nullopt;
+  }
+  auto dep_ext = resin::extension_to_shader_type(path.extension().string());
+
+  if (!dep_ext.has_value() || dep_ext.value() != ShaderType::Library) {
+    helpers::log_sh_err(curr_line_num, "The include macro argument must a library shader (.glsl extension).");
+    return std::nullopt;
+  }
+
+  return path;
+}
+
+static std::optional<std::string> process_ext_defi_macro(std::string_view arg, int64_t curr_line_num) {
+  if (!std::all_of(arg.begin(), arg.end(), [](const char c) { return std::isalnum(c) != 0 || c == '_'; })) {
+    helpers::log_sh_err(curr_line_num, "The external definition macro argument contains non-alphanumeric characters.");
+    return std::nullopt;
+  }
+
+  return std::string(arg);
+}
+
+ShaderResource ShaderResourceManager::load_res(const std::filesystem::path& path) {
+  auto sh_type = get_sh_type(path);
+  auto content = load_content(path);
+
+  auto lines = content | std::views::split('\n') | std::views::transform([](auto&& r) { return std::string_view(r); }) |
+               std::views::enumerate;
+
+  std::unordered_set<std::string> defi_names;
+  std::string preprocessed_content;
+
+  for (auto const [line, line_str] : lines) {
+    auto words = line_str | std::views::split(' ') |
+                 std::views::transform([](auto&& r) { return std::string_view(r); }) |
+                 std::views::filter([](auto chunk) { return !chunk.empty(); });
+
+    auto it  = words.begin();
+    auto end = words.end();
+
+    if (it == end) {
+      if (line_str != "") {
+        preprocessed_content.append(line_str);
+        preprocessed_content.append("\n");
+      }
+      continue;
+    }
+
+    auto macro = std::string_view{*it};
+    if (!shader_macros::is_macro(macro)) {
+      preprocessed_content.append(line_str);
+      preprocessed_content.append("\n");
+      continue;
+    }
+
+    ++it;
+    if (it == end) {
+      helpers::log_sh_err(line, "Macro {} requires an argument.", macro);
+      clear_and_throw<ShaderInvalidMacroArgumentsException>();
+    }
+    auto arg = std::string_view{*it};
+
+    ++it;
+    if (it != end) {
+      helpers::log_sh_err(line, "Macro {} expects exactly one argument.", macro);
+      clear_and_throw<ShaderInvalidMacroArgumentsException>();
+    }
+
+    if (macro == shader_macros::kIncludeMacro) {
+      auto rel_path = process_include_macro(arg, line);
+
+      if (!rel_path.has_value()) {
+        clear_and_throw<ShaderInvalidMacroArgumentsException>();
+      }
+
+      auto abs_path = path.parent_path() / rel_path.value();
+      if (std::ranges::find(visited_paths_, abs_path) != visited_paths_.end()) {
```

Comment:
_optional:_ additional check whether not including self, so we can save some time